README
/***********************************************************************
  Roster Master for Dragonfly(TM) CMS
  **********************************************************************
  Copyright (C) 2005-2020 by Dark Grue

  EverQuest II guild roster management, quest tracker, and dynamic
  signature generator that integrates with the Dragonfly(TM) Content
  Management System (CMS).

  Based on Roster Master by Rex "SaintPeter" Schrader.

  With gratitude for the support from:
    The Roster Master Project (http://www.rostermaster.org/),
    The EverQuest II Community (http://forums.daybreakgames.com/eq2/),
    The Dragonfly CMS Community (http://www.dragonflycms.org), and
    Roster Master users worldwide.

  License:
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or (at
  your option) any later version.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
  02111-1307 USA
***********************************************************************/


********************** Installation Requirements ***********************
Roster Master for Dragonfly (RM4DF) is a Dragonfly CMS module. For basic
installation, Dragonfly version 9.4 or greater is required. Other RM4DF
features may require additional dependencies, and are described below.


**************************** Module Install ****************************
This module is easy to install with no hassle from SQL files. It uses
Dragonfly's module management system to make it easier for everyone.

To install, just upload the content of this zip to your website "root"
directory so that the dirctories line up, like for like:

	blocks   => blocks
	language => language
	modules  => modules
	etc.

You may be warned: "Do you want to overwrite the folders?"... just say
yes; these won't replace or remove the wrong files.

In the "modules/Roster_Master/signatures",
"modules/Roster_Master/images/sigblocks", and "Roster_Master/fonts"
directories are "htaccess" files. The directives are not necessary for
the core operation of Roster Master, but can prevent wasted resources
and block external indexing from web spiders. Conversely, the .htaccess
file in /modules/Roster_Master provides additional functionality (URL
rewriting for signature blocks, primarily).

Edit modules/Roster_Master/config.inc to select the appropriate options
for your site. At a minimum, you MUST edit the values of the
$config['guild_id'] variable. If you do not know your Guild ID, there is
a search function in the Roster Master Administration menu to assist
you.

If you have chosen to use the logfile (by setting "$config['logfile']"
to anything other than zero), set the appropriate ownership and
permissions on the "modules/Roster_Master/logfile.inc" file. It is
recommended that it be owned/writable by the user the webserver runs
under. Some users who are running on hosted services may have to set
this file world-read/writable. Regardless of the situation, it is HIGHLY
recommended that this file be set to the most restrictive set of
permissions possible.

If you have chosen to use dynamic signatures (by setting
"$config['dynamic_sigs']" to "1"), ensure that the webserver process has
write access to the "modules/Roster_Master/signatures/" directory.

The Diagnostics page in the Roster Master Administration menu can be
extremely helpful in making sure that all module prerequisites are met.

Login as admin and go to the "Modules Administration" page, which will
autodetect the new module and check the cpg_inst.php file.

The cpg_inst.php file runs the necessary SQL commands to setup the tables
and make the appropriate changes to the database (select "Install").

When you see the module in the list of modules, check the "Active" box to
turn on the module to complete the installation.

That's all!

NOTE1: This module requires outbound HTTP connectivity to the
census.daybreakgames.com Website on TCP Port 80. Make sure your server
can connect outbound, or the module will be unable to function
(characterized by "FATAL: Gave up trying to contact the guild summary
page." and "FATAL Gave up trying to contact the roster page." messages in
the logfile). Fedora Core users (possibly others) will need to edit the
/etc/selinux/targeted/booleans.local file and add
"httpd_can_network_connect=1" (without the quotes). Other systems may
need similiar changes. Read the httpd_selinux(8) man page for more
information on this if your system uses this feature.

NOTE2: Do not use Dreamweaver to edit the Roster Master or Dragonfly PHP
files, there are known issues with Dreamweaver that can cause file
corruption and complicate troubleshooting. Use a plain-text editor (such
as Notepad or WordPad), or a programming editor, such as jEdit or
Notepad++.

NOTE3: The Guild Wall of Fame block stores a copy of its output in the
"content" field of the "{prefix}_blocks" table, and pulls from there
rather than re-run all the queries (until the block needs to be
updated). This works perfectly well until the stored block exceeds the
TEXT field type it's stored in which has a maximum length of 65,535
(2^16 - 1) characters. This condition would pretty much only exists if
$config['breakdown_shownames'] = 1, since the block stays a constant
length that's within the limits of the TEXT field. Also, people with
very small guilds, can possibly duck under the limit, even with it
enabled. To work around this, the Roster Master installer will change
the field type of the table from TEXT to MEDIUMTEXT. This should have no
impact on the operation of the Dragonfly CMS core, or to database
efficiency. However, please note that if you subsequently upgrade the
Dragonfly CMS core, there is a good chance that the CMS installer will
revert the field type back to TEXT, in which case it would have to be
remedied once again by hand (using whatever SQL management interface
that you are comfortable with). The relevant command is (make
appropriate substitutions):

	ALTER TABLE {prefix}_blocks CHANGE content content MEDIUMTEXT

If you need to disable this caching behavior in the block (useful for
testing), set the $DEBUG variable in the block-Guild_Wall_of_Fame.php
file to TRUE.


************************* Module Instantiation *************************
Multiple instances of this module can be created by making a copy of the
module directory and adding an underscore followed by an integer to the
name of the directory (e.g. "Roster_Master_2"). The module will detect
this and automagically configure itself.

NOTE: The module will check $mod_basename against its installation
directory. Blocks that reference this module MUST be configured
separately to reference the new instance (e.g. naming to
"block-Guild_Wall_of_Fame_2.php").
    
The naming of the module installation directory and block filenames are
set to follow a pattern. Both blocks and modules check against a
hardcoded base name and then look for an instantiation extension; the
module will NOT function if it does not pass that sanity check. The
correct naming convention is the base module installation directory
followed by an underscore and a number. Blocks follow the same pattern.
 
The module name may be changed from the default name of the installation
directory by going into the Module Administration Panel, selecting the
"Edit" option, and entering the desired title in the "Custom Title"
field, then pressing "Save Changes". Users using multiple instances will
almost certainly want to do this, to avoid unhelpful titles such as
"Roster Master", "Roster Master 2", "Roster Master 3", etc... Blocks can
be similiarly renamed from the Block Administration panel, selecting
"Edit", and entering the desired block title in the "Title" field, then
pressing "Save Changes".

NOTE: Don't rename the module's installation directory after you have
installed the module - neither the module nor the installer will be able
to find the module's tables, and you will be unable to uninstall the
module until it is renamed back to the original name the module was
installed under.


********************* JavaScript Scrolling Tables **********************
The Quest Master features (heritage, key, and timeline) offer by default
a JavaScript scrolling table and row highlighter in order to make
reading through the respective grids easier. These features can be
turned on or off by setting $config['qm_scrolltable'] and
$config['qm_rowhighlight'], in the config.inc file respectively.

The JavaScript re-writes the tracker pages dynamically, and is designed
to work with valid XHTML source. The script has been tested on IE,
Firefox, and Opera, but may display differently on other browsers, or
not work at all. If you experience problems, you may wish to check the
validity of your theme first to make sure it validates as XHTML. The
W3C Validator (http://validator.w3.org/) is the best method to check
your pages.


******************************** Avatars *******************************
Avatars are stored in the modules/Roster_Master/images/avatars
directory. In order to associate avatar images with a particular
character, place them in the
modules/Roster_Master/images/avatars/characterId subdirectory and name
them by the Character ID (e.g. "281167207.png").

The default avatar image for both the roster and signature blocks is
"anon.png". In order to use avatars, at least the "anon.png" file must be
present, as it is the file that the system falls back to if no other
appropriate one is available.

When the ARCH and RANK directives are in use, guild officers can be
differentiated with files prefixed by "officer_" (e.g.
"officer_priest.png"). If no officer avatar is available, it will fall
through to the member icon, and if the member icon is not available, it
will fall through to the default. Officers are indicated by populating
the $config['officers'] array with the RANK NUMBERS in the config.inc
file.

The ARCH directive selects one of four images, based on the character's
class archetype (Fighter, Mage, Priest, Scout). Filenames are the
archetype name in lowercase (e.g. "fighter.png).

The CLASS directive selects one of four images, based on the character's
class (Assassin, Berserker, Brigand, Bruiser, etc.). Filenames are the
class name in lowercase (e.g. "defiler.png).

The EQ2 directive selects a random icon (selections are not saved from
view to view) from the standard SoE EQ2 icons. Filenames are of the form
"eq2_*.png".

The RANK directive selects between two icons, "officer_anon.png" and
the default "anon.png".

Regardless of the avatar directive setting, Roster Master always shows
first preference to Character ID avatars.

Avatar images in the roster are hardcoded to display at 50 x 50 (set in
the config.inf file by the $config['avatar_size']['width'] and
$config['avatar_size']['height'] variables). As long as avatars are
square, they won't be distorted. However, keep in mind that if you are
using large avatars, the server is still transferring the entire image:
the browser is doing the scaling (so be mindful of your bandwidth).


********************* Dynamic Signature Generator **********************
Roster Master has the option of being able to generate dynamic signature
block images from the roster database. A caching system helps offload
resources from the PHP subsystem and the Web server. The cache works in
the following way:

* Registers a shutdown function to destory GD image resources.
* Gets the Character_ID from the server environment and casts it to
  integer.
* Gets the record matching the Character_ID from the roster, validating
  that it exists at the same time.
* Checks the server environment to see if a forced refresh has been
  requested by a CMS user, if it has, skip the caching logic.
* The caching logic starts by seeing if a signature block file already
  exists, gets the last modified time of the file, and checks that the
  file is at least as new as the latest information for that character
  in the roster. The roster is configured by default (see config.inc and
  the $config['cache_update_time'] parameter) to check Census data on
  the next roster or heritage quest tracker access after six hours.
* It then checks to see if the file modification time is greater than the
  If-Modified-Since header sent by the client. If it is, it pushes the
  cached file and exits. If it is not, it sends a 304 response and exits.
* If the caching logic has been skipped or exited, the module creates the
  image file, saves it to the cache directory, pushes it to the client,
  and exits.

This feature requires GD 2.0.1 or later and the FreeType library, both
of which are compile-time options for PHP. You must meet both of these
requirements in order to turn this feature on.

To enable the Signature Master feature, set the 
"$config['dynamic_sigs'] = 1" option in the
"modules/Roster_Master/config.inc" file.

You must also set the appropriate ownership and permissions on the
"modules/Roster_Master/signatures" directory. It is recommended that it be
owned/writable by the user the webserver runs under. Some users who are
running on hosted services may have to set this file world-read/writable.
Regardless of the situation, it is HIGHLY recommended that this file be set
to the most restrictive set of permissions possible.

Be aware that broswer caching performance/compatibility works best
(or at all) only with Dragonfly's Link Engine Optimization (LEO)
feature turned on. The difference is that without LEO, all signature
block links looks like:

	index.php?name=Roster_Master&rm=rm_sig&characterId=xxxxxxxxxx

Which really is just saying file "index.php". Without LEO on, your
entire Dragonfly site pretty much points to index.php, and most caches
will not be able to distinguish between the sifferent pages. That is,
the query/fragment portion of the URL is not (seldom) used as a
distinguisher in client caching schemes. With LEO on, a signature
block link looks like:

	Roster_Master/rm=rm_sig/characterId=xxxxxxxxxx.html

Which is unique, and therefore easier to distinguish as a cacheable
entity. Thus, it's recommended that LEO be used with the Signature
Master feature. (LEO in general is beneficial for your site, especially
if searchability by major indexing services like Google is important to
you.)

The signature block generator is set to generate 400 x 100 JPEG images,
suitable for use on the EQ2 forums, or other places. The maximum
allowed signature size on the EQ2 forums as of Live Update #29:
November 14, 2006 (Echoes of Faydwer) is 500 x 150 total, text and
images.

Block and avatar images may be GIF, JPEG, or PNG images. The module is
configured to get PNG images by default. Filenames may be changed by
editing the source of the
"modules/Roster_Master/functions/sigmaster.inc" file.

Base signature blocks need to be sized 400 x 100, and are stored in the
"modules/Roster_Master/images/sigblocks" directory.

Signature block directives are selected in the config.inc file by
setting "$config['sigblock_type']". The ARCH directive works much like
that for avatars. Blocks are named of the form
"<block basename>_officer_<archetype>.png" or
"<block basename>_<archetype>.png" (e.g.
"CFLogo_sigblock_officer_scout.png"). The CLASS directive replaces the
archetype with the class name, but otherwise uses the same naming
convention.

Signature blocks use the same avatar images as the roster. Avatar images
may be any size up to 100 x 100; smaller images will be centered in the
rightmost 100 x 100 corner of the base signature block.

Signature block templates are file-based. The template that Signature
Master will look for must be named the same as the signature block
image file (set by "$config['sigblock_file']" in config.inc and case-
sensitive), but with the "txt" extension. For example, if your
sigblock_file is named "CFLogo_sigblock.png", then your template file
should be named "CFLogo_sigblock.txt" and placed in the same directory.
If directives are in use, templates can be further broken down for each
sigblock image file, or one template may be used (named off the block
basename).

Sample files, "default_sigblock.txt" (the default) and
"CFLogo_sigblock.txt" have been provided for reference.


CHANGELOG
/***********************************************************************
  Roster Master for Dragonfly(TM) CMS
  **********************************************************************
  Copyright (C) 2005-2020 by Dark Grue

  EverQuest II guild roster management, quest tracker, and dynamic
  signature generator that integrates with the Dragonfly(TM) Content
  Management System (CMS).

  Based on Roster Master by Rex "SaintPeter" Schrader.

  With gratitude for the support from:
    The Roster Master Project (http://www.rostermaster.org/),
    The EverQuest II Community (http://forums.daybreakgames.com/eq2/),
    The Dragonfly CMS Community (http://www.dragonflycms.org), and
    Roster Master users worldwide.

  License:
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or (at
  your option) any later version.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
  02111-1307 USA
***********************************************************************/


Changelog:
* TODO: Check for new key quests. Would appreciate any feedback on useful quests to include here (such as relevant raid flags).
* TODO: Migrate remaining administrative functions to panel (maybe - they might be more convenient where they are).

9.1.2
* Added new collections for Blood of Luclin.
* Created handler for missing collection icons.
* Removed "elements parsed." message from the quest array generator, as it
  was screwing up cut-and-paste of the data into the include files.

9.1.1
* Corrected version number in file headers.
* Made some changes to parser logic to make it less fragile to anomalies in the
  Census data, thanks to Demetrios for finding the issue.
* Fixed some spelling errors in comments.
* Added sort of collection quest by number collected.
* Improved highlighting style of missing collection objects.
* Added count of collection quests that have not been started.
* Added indicator for collection quest class-restricted and meta-collections.
* Added new collections for Seeds of Vengeance (GU106).
* Added new collections for Return to Guk (GU107).
* Added new collections for Chaos Descending (GU108).

9.1.0
* Improved Alt logic.
* Added the ability to show only claimed characters in Quest Master. Thanks to
  Tulvarus, for the feature suggestion.
* Fixed bug in Quest Master where resort links weren't preserving Expansion
  selections.
* Query Quest Status now honors the current show_alts and show claimed
  characters search options.
* Added vertical centering to sort checkboxes and radio button labels.
* Replaced references to http://forums.station.sony.com/eq2/ with
  http://forums.daybreakgames.com/eq2/ link.
* Fixed database errors with references to "rm.id".
* Fixed problem with multiple crc entries for the same characterID being
  present in the roster_master_quest_status table.
* Fixed old RM4DF version number in sigblock templates.
* Fixed Signature Master not pulling signature block avatars from the correct
  subdirectories and defaulting to the eq2 logos.
* Some template cleanup.
* Moved administration display strings into language file.
* Additional diagnostics messages.
* Cleaned up installer's database upgrade strategy.
* Remove obsolete statistics from block.
* Updated documentation.
* Updated contib/history.sql file to reflect new events.
* New feature: addition of Collection Master!
  + Reports on a character's collection quests.
  + Like Quest Master, automatically updated from Census.
  + Note: Collection Master can only know about collection quests that exist
    in Census, quests have been known to be held back for new expansions.
* Reworked parser to increase efficiency:
  + Reduced use of globals and reworked data structures to be more efficient
    for write operations, rather than replicate the structures used to build
	the roster pages.
  + SQL query optimizations.
  + Massive speed improvements during roster update; NOTE: this comes at the
    cost of detailed logging of roster changes (eliminating the loops to do
	comparisons solely for logging purposes had a significant performance
	cost).
  + Removed XML parsers in favor of JSON, for a 25% data savings advantage.
* Thanks to Feldon with helping with the Census query syntax and suggestions
  on effeciency improvements.

9.0.0
* Fixed sitewide "Accessed Denied" error when module is uninstalled.
* Added missing Aerakyn race icons.
* Added missing Channeler class icon and fixed missing display logic.
* Getting ready for 2038, changed DB time representations to BIGINT.
* Added tooltips to sort options to clarify interface.
* Minor cleanup to SQL query syntax.
* Deprecated use of id field for character data (characterId is now
  canonical).
* New feature: major update to Quest Master!
  + Quests automatically updated from Census, no need to manually
    set quest status!!!!
  + Adding new quests doesn't require DB schema changes.
  + Updated sort dialogs with tooltips for abbreviated items.
  + Note that extensive internal changes have been made, files have been
    added and removed.
  + Better handling of sort options.
* Rework to XML parsers.
* Removed $config['qm_exclude_inactive'] option.
* Fixed table resize JavaScript functionality that was preventing the feature
  from working.
* Sort direction indicators are now clickable to change sort direction for
  that column.
* Removed obsolete access quests (tSS, NekC_tR, CoK, tSC).
* Added Census service ID for the Roster Master application. Please refer to

language/english/roster_master.php
<?php
/***********************************************************************
  Roster Master for Dragonfly(TM) CMS
  **********************************************************************
  Copyright (C) 2005-2020 by Dark Grue

  EverQuest II guild roster management, quest tracker, and dynamic
  signature generator that integrates with the Dragonfly(TM) Content
  Management System (CMS).

  License:
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or (at
  your option) any later version.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
  02111-1307 USA
***********************************************************************/
if (!defined('CPG_NUKE')) { exit; }


// Marker
define('_ROSTERMASTER8', 'English');

// _ACTIVE
// _ADMIN
// _ADMINISTRATION
define('_ADORNING', 'Adorning');
define('_ADORNINGT', 'Adrn');
define('_ADVCLASS', 'Adventurer Class');
define('_ADVCLASST', 'Adv');
define('_ADVLEVEL', 'Adventurer Level');
define('_ADVLEVELT', 'Lvl');
define('_ALTS', 'Alts:');
define('_ARTCLASS', 'Artisan Class');
define('_ARTCLASST', 'Art');
define('_ARTLEVEL', 'Artisan Level');
define('_ARTLEVELT', 'Lvl');
define('_ASC', 'A'); // ascending
define('_ASCENDING', 'Ascending');
define('_AVATAR_RM', 'Avatar');
define('_BBCODE', 'BBCode');
define('_BUILDCOLLECTIONS', 'Build Collections Data');
define('_BUILDCOLLECTIONSDESC', 'Rebuild the $CollectionsData array from the Census.');
define('_BUILDICONS', 'Build Item Icon Cache');
define('_BUILDICONSDESC', 'Download and save to disk all unique collection icons.');
define('_BUILDQUESTS', 'Build Quests Data');
define('_BUILDQUESTSDESC', 'Rebuild the $QuestsData array from the Census.');
// _BC_DELIM
// _CATEGORY
define('_CENSUS', 'Census');
define('_CHARACTER', 'Character');
define('_CHARACTERID', 'Character ID');
define('_CHARACTERTYPE', 'Character Type');
define('_CHARACTERTYPET', 'Type');
define('_CMSVERSION', 'CMS Version');
define('_CLASS', 'CLASS');
define('_COLLECTABLE', 'Collectable');
define('_COLLECTIONS', 'Collections');
define('_COMPLETED', 'Completed');
define('_COUNT', 'Count');
define('_CURRENTCLAIM', 'Current Claims');
define('_DELCLAIM', 'Delete Claim');
define('_DESC', 'D'); // descending
define('_DESCENDING', 'Descending');
define('_DEVELOPERTOOLS', 'Developer Tools');
define('_DIAGNOSTICS', 'Diagnostics');
define('_DIAGNOSTICSDESC', 'Perform system configuration diagnostics relevant to the Roster Master module.');
define('_DISPALL', 'Display All');
define('_DISPPRI', 'Display Primary');
define('_DISPLAYING', 'Displaying');
define('_DONTHAVE', 'Not Started');
define('_EDITCLAIM', 'Edit Claim');
define('_EDITCHARCLAIM', 'Edit Character Claim');
define('_END', 'End');
define('_EXPANSION', 'Expansion');
define('_FILLSIGCACHE', 'Fill Signature Block Cache');
define('_FILLSIGCACHECOMP', 'Signature block caching operation completed.');
define('_FILLSIGCACHECONF', 'Do you really wish to prepopulate all cached signature block images? This can take an extended amount of time (and exceed the time a script is allowed to run).');
define('_FILLSIGCACHEDESC', 'Fill the signature block cache.');
define('_FORCEUPDATE', 'Force Update');
define('_GROUPED', 'Grouped');
define('_GUILDBANK', 'Guild Bank');
define('_GUILDLEVEL', 'Guild Level');
define('_GUILDNAME', 'Guild Name');
define('_GUILDRANK', 'Guild Rank');
define('_GUILDRANKT', 'Rank');
define('_HAVESTARTER', 'Have Starter');
define('_HIGHESTMAGICHIT', 'Max Magic');
define('_HIGHESTMELEEHIT', 'Max Melee');
define('_HTML', 'HTML');
define('_INPROGRESS', 'In Progress');
define('_JOINEDON', 'Joined');
define('_KVDRATIO', 'KVD Ratio');
define('_LASTNAME', 'Lastname');
define('_LASTONLINE', 'Online');
define('_LASTUPDATE', 'Last Updated');
define('_LEGEND', 'Legend');
define('_LEVEL', 'Level');
define('_LINKTOCACHED', 'Link to Cached Image');
define('_LINKTOCHAR', 'Link Image to Character');
define('_LOG', 'Log');
define('_LOGPRGECOMP', 'Logfile purge operation completed.');
define('_LOOKUPGID', 'Lookup Guild ID');
define('_LOOKUPGIDDESC', 'Find a guild\'s unique Guild ID by searching guild names and servers. (<a href="http://en.wikipedia.org/wiki/Regular_expression" target="_blank">Regular expression</a> syntax allowed.)');
define('_LOOKUPGIDTRUNC', '%d results found - displaying first %d only.<br /><br />If the results you are looking for do not appear in this list, narrow your search by making it more specific.<br /><b>TIP:</b> Anchoring the start of your search string by prefixing it with a "^" to anchor the start of the search and/or "$" to anchor the end will usually produce the desired results.');
define('_MANAGECHAR', 'Manage Characters');
define('_MANAGEROSTER', 'Manage Roster');
define('_MEMBERCOUNT', 'Members');
define('_META', 'META');
define('_NEWCLAIM', 'New Claim');
// _NAME
define('_NAMET', 'Name');
// _NO
// _NONE
define('_OF', 'of');
define('_OPCANCEL', 'Operation CANCELLED by user request.');
define('_OPTIMZETABLES', 'Optimize Tables');
define('_OPTIMIZETABLESDESC', 'Optimize the database tables by removing all widowed and orphaned records from the database.');
define('_OPTIMIZETABLESCONF', 'Do you wish to delete all widowed and orphaned claims? <b>NOTE:</b> This will delete the character claim, all quest status, and dynamic signature configurations for all affected records.');
define('_OPTIMIZETABLESCOMP', 'Table optimization completed.');
define('_ORPHANCLAIM', 'Orphaned Claims');
define('_PHPDIAGNOSTICSDESC', 'Display core PHP configuration for PHP.');
define('_PHPINFO', 'PHP Info');
define('_PHPVERSION', 'PHP Version');
define('_PRIMARY', 'Primary');
define('_PROGRESS', 'Progress');
define('_PURGELOGDESC', 'Empty the log file of all records.');
define('_PURGELOGCONF', 'Do you wish to purge the logfile of all data?');
define('_PURGELOGFILE', 'Purge Logfile');
define('_PURGESIGCACHE', 'Purge Signature Block Cache');
define('_PURGESIGCACHEDESC', 'Purge the signature block cache.');
define('_PURGESIGCACHECOMP', 'Signature block cache purge operation completed.<br /><br />Removed %d signature files from cache.');
define('_PURGESIGCACHECONF', 'Do you really wish to remove all cached signature block images?');
define('_PURGETABLE', 'Purge Roster Table');
define('_PURGETABLECONF', 'Do you wish to purge the Roster Table of all parsed roster data? <b>NOTE:</b> This will <i>not</i> affect Claims. If you need to zero out all Roster Master data, uninstall, then reinstall the module from the <a href="%s">'._MODULESADMIN.'</a> panel.');
define('_PURGETABLECOMP', 'Purge table operation completed.');
define('_PURGETABLEDESC', 'Purge parsed Roster Master data from database tables.');
define('_QUERY', 'Query');
define('_QUERYQS', 'Query Quest Status');
define('_QUEST', 'Quest');
define('_QUESTS', 'Quests');
define('_QUESTST', 'Quests');
define('_QUESTNAME', 'Quest Name');
define('_RACE', 'Race');
define('_RACET', 'Race');
define('_RECLVL', 'Recommended Level');
define('_RECLVLT', 'Rec Lvl');
define('_REQLVL', 'Required Level');
define('_REQLVLT', 'Req Lvl');
define('_REWARD', 'Reward');
define('_ROSTER', 'Roster');
define('_SAVECLAIM', 'Save Claim');
// _SEARCH
define('_SERVER', 'Server');
define('_SIGIMG', 'Signature Block Image');
define('_SIGLINK', 'Signature Block Link');
define('_SIGPREVIEW', 'Signature Block Preview');
define('_SITEROOT', 'Path from site root');
define('_SHOWALL', 'Showing all characters.');
define('_SHOWPRI', 'Showing Primary characters only.');
define('_SHOWCO', 'Show your Claimed characters only');
define('_SHOWCQ', 'Show completed quests');
define('_SHOWIQ', 'Show inactive quests');
define('_SORT', 'Sort');
define('_SORTPRI', 'Primary Sort');
define('_SORTSEC', 'Secondary Sort');
define('_STAGE', 'Stage');
define('_STARTER', 'Starter?');
define('_STATUSPOINTS', 'Status');
define('_STATUSPOINTST', 'Status');
define('_STEP', 'Step');
define('_SYNCCLAIMS', 'Sync Active Claims');
define('_TINKERING', 'Tinkering');
define('_TINKERINGT', 'Tink');
define('_TOTAL', 'Total');
define('_TOTALMEM', 'Total Members');
define('_TRACKERCOLLECT', 'Collection Quest Tracker');
define('_TRACKERHERITAGE', 'Heritage Quest Tracker');
define('_TRACKERKEY', 'Access Quest Tracker');
define('_TRACKERTIMELINE', 'Timeline Quest Tracker');
// _TYPE
define('_UNCLAIMEDCHARS', 'Unclaimed Characters');
define('_UNGROUPED', 'Ungrouped');
define('_UNKNOWN_RM', 'Unknown');
define('_USERID_RM', 'User ID');
define('_USERNAME_RM', 'Username');
define('_USERNAMET', 'User');
define('_VIEWLOG', 'View the Roster Master module log.');
define('_VIEWLOGFILE', 'View Logfile');
define('_VIEWROSTER', 'Guild Roster');
define('_WIDOWCLAIM', 'Widowed Claims');
// _YES

define('_LOG_NOLOGFILE', 'No Logfile Found.');
define('_LOG_NOLOGCONTENT', 'Log is empty.');
define('_LOG_PROCESSCOMPLETE', 'Parsed %d (of %d total) character record(s) into the roster database. Removed %d characters. Deleted %d table rows from removed characters.');
define('_LOG_ROSTERADD', 'Roster change: %s added to local database.');
define('_LOG_ROSTERDEL', 'Roster change: %s removed from local database.');
define('_LOG_ROSTERFIELDCHANGE', '%s: %s was %s, is now %s.');
define('_LOG_SUMMARYCHANGE', 'Guild Summary: %s was %s, is now %s.');
define('_LOG_UPDATE', 'Update');

define('_NTC_GETJSON', 'Time to retrieve raw JSON %s collection (%s): %.4f sec.');
define('_NTC_JSONPARSE', 'Time to parse JSON %s collection: %.4f sec, time to process: %.4f sec, total parser time: %.4f sec.');
define('_NTC_RSTRDISPLAY', 'Time to display roster (%s): %.4f sec.');
define('_NTC_SMCFILL', 'Time to populate signature block cache (%d of %d images): %.4f sec.');
define('_NTC_UPDATEDATA', 'Time to update Guild Summary: %.4f sec total (+%.4f sec for parse, +%.4f for db write, +%.4f sec processing);<br />
Character Collection: %.4f sec total (+%.4f sec for parse, +%.4f for db write [prep statements: %.4f sec, delete tables: %.4f sec, write roster %.4f sec, write collections %.4f sec, write quests %.4f sec], +%.4f sec processing);<br />
total update time: %.4f sec.');

define('_ERR_ACCESSCONT', 'FATAL: Access control error. You are not permitted to complete this operation.');
define('_ERR_CLAIMNONE', 'No characters claimed.');
define('_ERR_CLAIMREADFAIL', 'FATAL: Unable to retrieve claim from database.');
define('_ERR_CLAIMSAVEFAIL', 'FATAL: Unable to add claim to database.');
define('_ERR_CSVFAIL', 'CSV Parse Failure: no matches found.');
define('_ERR_CURLINFO_SIZE_DOWNLOAD', 'cURL downloaded %s (raw), %s (uncompressed); %.2f%% savings due to compression.');
define('_ERR_CURLERR', 'FATAL: cURL error "%s" returned.');
define('_ERR_CURLNOTFOUND', 'WARNING: cURL support not found. Falling back to internal method.');
define('_ERR_DBCE', 'Database consistency error, more than one result returned.');
define('_ERR_FATALCONFIG_RM', 'FATAL: A configuration error has occurred.');
define('_ERR_FILENOTFOUND', 'FATAL: File "%s" not found.');
define('_ERR_FILEHANLEFAIL', 'FAIL: Failed to fopen() file handle for file "%s".');
define('_ERR_GUILDNOTFOUND', 'No guild found!');
define('_ERR_HTTPRESPONSEFAIL', 'FATAL: Recieved HTTP response status code other than success or redirection. Actual code recieved was %s. Error may indicate URL path (%s) is faulty.');
define('_ERR_INSTANTIATION', 'The module installation directory, <nobr>"%s"</nobr>, does not follow the expected instantation syntax of <nobr>"%s[_<i>&lt;integer&gt;</i>]"</nobr>.');
define('_ERR_JSON_FAIL1', 'WARNING: fetch_url() unable to contact the JSON collection URL, Sleeping %s seconds...');
define('_ERR_JSON_FAIL2a', 'FATAL: Gave up trying to contact the JSON collection URL.');
define('_ERR_JSON_FAIL2b', 'Try refreshing this page or visiting %s directly.');
define('_ERR_PARSE', 'FATAL: There was an error during the parse operation.');
define('_ERR_LOGWRITE', 'Unable to open logfile.inc for writing.<br />Make sure that the webserver has write permission set for this file AND the install folder.');
define('_ERR_NOSERVICEID', 'WARNING: Missing Service ID. Rate limited to less than 10 queries per minute. Retrying in 60 seconds...');
define('_ERR_PROCESSFAIL', 'FATAL: %s parser processing failure. Roster database NOT updated.');
define('_ERR_ROSTEREXPIRED', 'Roster data cache expired. Data stale for %s');
define('_ERR_STREAMBLOCKING', 'NOTICE: Unable to set stream to non-blocking mode. (Safe to ignore.)');
define('_ERR_STREAMTIMEOUT', 'WARNING: Stream timed out!');
define('_ERR_URL', 'Fetching URL: ');
define('_ERR_URLREDIRECT', 'REDIRECT to URL: ');

define('_ERR_XML_FAIL1', 'WARNING: fetch_url() unable to contact the XML collection page, Sleeping %s seconds...');
define('_ERR_XML_FAIL2a', 'FATAL: Gave up trying to contact the XML collection page.');
define('_ERR_XML_FAIL2b', 'Try refreshing this page or visiting %s directly.');
